diff --git a/include/wifi_base.h b/include/wifi_base.h
index b7e4d3a57..a9c46481b 100644
--- a/include/wifi_base.h
+++ b/include/wifi_base.h
@@ -38,7 +38,9 @@ extern "C" {
 #define WIFI_STA_INTERFACE_NAME             "Device.WiFi.STA.{i}.InterfaceName"
 #define WIFI_STA_CONNECTED_GW_BSSID         "Device.WiFi.STA.{i}.Bssid"
 #define WIFI_STA_SELFHEAL_CONNECTION_TIMEOUT "Device.WiFi.STAConnectionTimeout"
+#define WIFI_ENDPOINT_CONNECT_STATUS        "Device.WiFi.EndPoint.1.Status"
 #define WIFI_ACTIVE_GATEWAY_CHECK           "Device.X_RDK_GatewayManagement.ExternalGatewayPresent"
+#define RF_STATUS_CHECK                     "Device.WiFi.EndPoint.1.Enable"
 #define WIFI_WAN_FAILOVER_TEST              "Device.WiFi.WanFailoverTest"
 #define WIFI_LMLITE_NOTIFY                  "Device.Hosts.X_RDKCENTRAL-COM_LMHost_Sync_From_WiFi"
 #define WIFI_HOTSPOT_NOTIFY                 "Device.X_COMCAST-COM_GRE.Hotspot.ClientChange"
diff --git a/source/core/services/vap_svc.h b/source/core/services/vap_svc.h
index 9fed6d316..1083ced24 100644
--- a/source/core/services/vap_svc.h
+++ b/source/core/services/vap_svc.h
@@ -65,7 +65,7 @@ typedef bool (* vap_svc_is_my_fn_t)(unsigned int vap_index);
 
 #define EXT_SCAN_RESULT_TIMEOUT                4000
 #define EXT_SCAN_RESULT_WAIT_TIMEOUT           4000
-#define EXT_CONN_STATUS_IND_TIMEOUT            5000
+#define EXT_CONN_STATUS_IND_TIMEOUT            12000
 #define EXT_CSA_WAIT_TIMEOUT                   3000
 #define EXT_DISCONNECTION_IND_TIMEOUT          5000
 #define EXT_UDHCP_IP_CHECK_INTERVAL            60000
diff --git a/source/core/services/vap_svc_mesh_ext.c b/source/core/services/vap_svc_mesh_ext.c
index 143c283d6..82de9106e 100644
--- a/source/core/services/vap_svc_mesh_ext.c
+++ b/source/core/services/vap_svc_mesh_ext.c
@@ -31,7 +31,9 @@
 #include "wifi_ctrl.h"
 #include "wifi_mgr.h"
 #include "wifi_util.h"
+#include "wifi_stubs.h"
 #include "wifi_hal_rdk_framework.h"
+#include "wifi_base.h"
 
 #define PATH_TO_RSSI_NORMALIZER_FILE "/tmp/rssi_normalizer_2_4.cfg"
 #define DEFAULT_RSSI_NORMALIZER_2_4_VALUE 20
@@ -40,6 +42,9 @@
 #define EXT_DISCONNECTION_DISCONNECT 1
 #define EXT_DISCONNECTION_DISCONNECT_AND_IGNORE_RADIO 2
 
+int scan_trigger_counter;
+int scan_trigger_threshold = 10;
+
 static void swap_bss(bss_candidate_t *a, bss_candidate_t *b)
 {
     bss_candidate_t t = *a;
@@ -332,7 +337,7 @@ int process_disconnection_event_timeout(vap_svc_t *svc)
 
     return 0;
 }
-
+#if 0
 static int process_udhcp_disconnect_event_timeout(vap_svc_t *svc)
 {
     wifi_ctrl_t *ctrl = svc->ctrl;
@@ -374,7 +379,7 @@ static int process_udhcp_disconnect_event_timeout(vap_svc_t *svc)
 
     return 0;
 }
-
+#endif
 static int process_trigger_disconnection_event_timeout(vap_svc_t *svc)
 {
     wifi_ctrl_t *ctrl = svc->ctrl;
@@ -418,6 +423,7 @@ static int process_trigger_disconnection_event_timeout(vap_svc_t *svc)
 
 int process_udhcp_ip_check(vap_svc_t *svc)
 {
+#if 0
     static int ip_check_count = 0;
     struct sockaddr_in sa;
     char value[128];
@@ -489,6 +495,7 @@ int process_udhcp_ip_check(vap_svc_t *svc)
     }
 
     ip_check_count++;
+#endif
     return 0;
 }
 
@@ -590,10 +597,24 @@ void ext_process_scan_list(vap_svc_t *svc)
         if (ext->candidates_list.scan_count != 0) {
             ext_set_conn_state(ext, connection_state_connection_in_progress, __func__, __LINE__);
         } else {
-            ext_set_conn_state(ext, connection_state_disconnected_scan_list_none, __func__,
-                __LINE__);
+#if 0
+	     int ret = RETURN_OK;	
+	     scan_trigger_counter++;
+             wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD] scan_trigger_counter : %d scan_trigger_threshold : %d\n", __func__, __LINE__, scan_trigger_counter, scan_trigger_threshold);
+             if (scan_trigger_counter >= scan_trigger_threshold) {
+                 wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD] scan_trigger_counter : %d scan_trigger_threshold : %d\n", __func__, __LINE__, scan_trigger_counter, scan_trigger_threshold);
+                 int connection_status = 2; //Sta disconnected
+		 ret = publish_endpoint_status_to_wan(ctrl, connection_status);
+                 if (ret == RETURN_ERR) {
+                      wifi_util_dbg_print(WIFI_CTRL,"%s:%d Error in publishing the status\n");
+                 }
+		 return;
+              } else {
+                    wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD] scan_trigger_counter : %d scan_trigger_threshold : %d\n", __func__, __LINE__, scan_trigger_counter, scan_trigger_threshold);
+#endif
+                    ext_set_conn_state(ext, connection_state_disconnected_scan_list_none, __func__,
+                        __LINE__);
         }
-
         schedule_connect_sm(svc);
     } else {
         wifi_util_dbg_print(WIFI_CTRL,"%s:%d wifi connection already in process state\n",__func__, __LINE__);
@@ -805,7 +826,7 @@ void ext_try_connecting(vap_svc_t *svc)
             wifi_util_error_print(WIFI_CTRL, "%s:%d sta connect failed for vap index: %d, "
                 "retry after timeout\n", __func__, __LINE__, vap_index);
         }
-
+        wifi_util_info_print(WIFI_CTRL,"%s:%d [PRAMOD]\n", __func__, __LINE__);
         if (ext->ext_conn_status_ind_timeout_handler_id != 0) {
             wifi_util_dbg_print(WIFI_CTRL, "%s:%d connect status timer is in progress, cancel\n",
                 __func__, __LINE__);
@@ -1090,6 +1111,7 @@ int vap_svc_mesh_ext_update(vap_svc_t *svc, unsigned int radio_index, wifi_vap_i
 {
     unsigned int i;
     wifi_vap_info_map_t tgt_vap_map;
+    vap_svc_ext_t *ext = &svc->u.ext;
 
     for (i = 0; i < map->num_vaps; i++) {
         memset((unsigned char *)&tgt_vap_map, 0, sizeof(tgt_vap_map));
@@ -1117,8 +1139,18 @@ int vap_svc_mesh_ext_update(vap_svc_t *svc, unsigned int radio_index, wifi_vap_i
             &rdk_vap_info[i]);
         get_wifidb_obj()->desc.update_wifi_security_config_fn(getVAPName(map->vap_array[i].vap_index),
             &map->vap_array[i].u.sta_info.security);
-    }
+        
+	if (!is_sta_enabled()) {
+            ext_set_conn_state(ext, connection_state_disconnected_steady, __func__, __LINE__);
+        } else {
+            ext_set_conn_state(ext, connection_state_disconnected_scan_list_none, __func__, __LINE__);
+            wifi_util_info_print(WIFI_CTRL, "%s:%d sta is enabled starting the station vaps\n",__FUNCTION__,__LINE__);
+            schedule_connect_sm(svc);
 
+            ext->is_started = true;
+
+        }
+    }
     return 0;
 }
 
@@ -1485,6 +1517,61 @@ static int apply_pending_channel_change(vap_svc_t *svc, int vap_index)
     return RETURN_OK;
 }
 
+int get_endpoint_enable(wifi_ctrl_t *ctrl)
+{
+    int rc = bus_error_success;
+    raw_data_t data;
+    memset(&data, 0, sizeof(raw_data_t));
+    bool endpoint_enable;
+    rc = get_bus_descriptor()->bus_data_get_fn(&ctrl->handle, RF_STATUS_CHECK, &data);
+    if (data.data_type != bus_data_type_boolean) {
+        wifi_util_error_print(WIFI_CTRL,"%s:%d %s bus_data_get_fn failed with data_type:0x%x, rc:%d\n", __func__, __LINE__, RF_STATUS_CHECK, data.data_type, rc);
+   }
+
+    if (rc != bus_error_success) {
+         wifi_util_error_print(WIFI_CTRL, "%s:%d bus_geti_fn failed for [%s] with error [%d]\n",
+           __func__, __LINE__, ETH_BH_STATUS, rc);
+    }
+    endpoint_enable = data.raw_data.b;
+    wifi_util_dbg_print(WIFI_CTRL, "%s:%d: event: %s value: %d\n", __func__, __LINE__, RF_STATUS_CHECK, endpoint_enable);
+    return endpoint_enable;
+}
+
+#define MAX_STATUS_LEN 5
+
+int publish_endpoint_status_to_wan(wifi_ctrl_t *ctrl, int connection_status) {
+    int ret = 0;
+    char name[128] = {'\0'};
+    bus_error_t rc = bus_error_success;
+    wifi_util_info_print(WIFI_CTRL,"%s:%d Conn-status updated as %d\n", __func__, __LINE__, connection_status);	
+    ret = get_endpoint_enable(ctrl);
+    wifi_util_info_print(WIFI_CTRL,"%s:%d ret updated as %d\n", __func__, __LINE__, ret);
+    if (ret == true) {
+	raw_data_t data;
+        sprintf(name, "Device.WiFi.EndPoint.1.Status");
+	memset(&data, 0, sizeof(raw_data_t));
+        data.data_type = bus_data_type_string;
+        data.raw_data.bytes = malloc(MAX_STATUS_LEN);
+        data.raw_data_len = MAX_STATUS_LEN;
+        memset(data.raw_data.bytes, '\0', MAX_STATUS_LEN);
+	if (connection_status == 2) { // connected state
+            strncpy((char *)data.raw_data.bytes, "Up", MAX_STATUS_LEN);
+	} else if (connection_status ==1) { //discpnnected state
+            strncpy((char *)data.raw_data.bytes, "Down", MAX_STATUS_LEN);
+	}
+	rc = get_bus_descriptor()->bus_event_publish_fn(&ctrl->handle, name,  &data);
+        if (rc != bus_error_success) {
+            wifi_util_dbg_print(WIFI_CTRL, "%s:%d: bus_event_publish_fn(): Event failed\n", __func__, __LINE__);
+            return RETURN_ERR;
+        }
+    } else {
+	wifi_util_info_print(WIFI_CTRL,"%s:%d Endpoint not enabled\n", __func__, __LINE__);
+	return RETURN_OK;
+    }
+    return RETURN_OK;
+}
+
+
 int process_ext_sta_conn_status(vap_svc_t *svc, void *arg)
 {
     wifi_mgr_t *mgr = (wifi_mgr_t *)get_wifimgr_obj();
@@ -1497,11 +1584,14 @@ int process_ext_sta_conn_status(vap_svc_t *svc, void *arg)
     bool found_candidate = false, send_event = false;
     unsigned int i = 0, index, j = 0;
     char name[64];
+    char cmd[128] = {0};
     wifi_sta_conn_info_t sta_conn_info;
     wifi_radio_operationParam_t *radio_params = NULL;
     wifi_radio_feature_param_t *radio_feat = NULL;
     bus_error_t rc;
     raw_data_t data;
+    char *bridge_name = "brww0";
+    int ret = 0;
 
     ctrl = svc->ctrl;
     ext = &svc->u.ext;
@@ -1519,10 +1609,31 @@ int process_ext_sta_conn_status(vap_svc_t *svc, void *arg)
         ext_conn_status_to_str(sta_data->stats.connect_status));
     vap_map = &mgr->radio_config[index].vaps.vap_map;
 
+#if 0 //DL
+    if (sta_data->stats.connect_status == 1) {
+        ret = get_endpoint_enable(ctrl);
+	wifi_util_info_print(WIFI_CTRL,"%s:%d ret updated as %d\n", __func__, __LINE__, ret);
+	if (ret == true) {
+	    sprintf(name, "Device.WiFi.EndPoint.1.Status");
+	    uint32_t str_size = strlen("Down") + 1;
+	    memset(&data, 0, sizeof(raw_data_t));
+            data.data_type = bus_data_type_string;
+            data.raw_data.bytes = malloc(str_size);
+            data.raw_data_len = str_size;
+            strncpy((char *)data.raw_data.bytes, "Down", str_size);
+	    rc = get_bus_descriptor()->bus_event_publish_fn(&ctrl->handle, name,  &data);
+            if (rc != bus_error_success) {
+                wifi_util_dbg_print(WIFI_CTRL, "%s:%d: bus_event_publish_fn(): Event failed\n", __func__, __LINE__);
+    //            return RETURN_ERR;
+            }
+	}
+    }
+#endif
 
     for (i = 0; i < vap_map->num_vaps; i++) {
         if (vap_map->vap_array[i].vap_index == sta_data->stats.vap_index) {
             if (vap_map->vap_array[i].u.sta_info.conn_status != sta_data->stats.connect_status) {
+                wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD]\n", __func__, __LINE__);
                 // send bus connect indication
                 send_event = true;
             }
@@ -1534,12 +1645,14 @@ int process_ext_sta_conn_status(vap_svc_t *svc, void *arg)
             }
             temp_vap_info->u.sta_info.conn_status = sta_data->stats.connect_status;
             memset(temp_vap_info->u.sta_info.bssid, 0, sizeof(bssid_t));
+                wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD]\n", __func__, __LINE__);
 
             // Avoid releasing IP on connected interface when disconnect is received on another.
             // Mesh agent supports one DHCP client for all sta interfaces so when it receives any
             // disconnection event it will release IP.
             if (temp_vap_info->u.sta_info.conn_status != wifi_connection_status_connected &&
                     temp_vap_info->vap_index != ext->connected_vap_index) {
+                wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD]\n", __func__, __LINE__);
                 send_event = false;
             }
             break;
@@ -1551,11 +1664,48 @@ int process_ext_sta_conn_status(vap_svc_t *svc, void *arg)
         return RETURN_ERR;
     }
 
+    if (send_event == true) {
+        wifi_util_dbg_print(WIFI_CTRL,"Now connected %s:%d\n",__func__,__LINE__);
+#if 0
+        wifi_hal_add_station_bridge(sta_data->interface_name,bridge_name);
+
+        snprintf(cmd, sizeof(cmd), "ip link set dev %s up", bridge_name);
+        wifi_util_dbg_print(WIFI_CTRL,"%s:%d cmd : %s\n",__func__,__LINE__, cmd);
+        get_stubs_descriptor()->v_secure_system_fn(cmd);
+        wifi_util_dbg_print(WIFI_CTRL,"Now connected %s:%d\n",__func__,__LINE__);
+#endif
+        sprintf(name, "Device.WiFi.STA.%d.Connection.Status", index + 1);
+
+        wifi_util_dbg_print(WIFI_CTRL, "%s:%d bus name: %s connection status: %s\n", __func__,
+            __LINE__, name, ext_conn_status_to_str(sta_data->stats.connect_status));
+
+        memset(&sta_conn_info, 0, sizeof(wifi_sta_conn_info_t));
+
+        sta_conn_info.connect_status =  sta_data->stats.connect_status;
+        memcpy(sta_conn_info.bssid, sta_data->bss_info.bssid, sizeof(sta_conn_info.bssid));
+
+        memset(&data, 0, sizeof(raw_data_t));
+        data.data_type = bus_data_type_bytes;
+        data.raw_data.bytes = (void *)&sta_conn_info;
+        data.raw_data_len = sizeof(wifi_sta_conn_info_t);
+
+        rc = get_bus_descriptor()->bus_event_publish_fn(&ctrl->handle, name, &data);
+        if (rc != bus_error_success) {
+            wifi_util_dbg_print(WIFI_CTRL, "%s:%d: bus_event_publish_fn(): Event failed\n", __func__, __LINE__);
+            return RETURN_ERR;
+        }
+
+        wifi_util_dbg_print(WIFI_CTRL, "%s:%d interface_name=%s\n", __func__, __LINE__,sta_data->interface_name);
+
+   }
+
     if (sta_data->stats.connect_status == wifi_connection_status_connected) {
         if ((ext->conn_state == connection_state_connection_in_progress) ||
             (ext->conn_state == connection_state_connection_to_lcb_in_progress) ||
             (ext->conn_state == connection_state_connection_to_nb_in_progress)) {
+                wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD]\n", __func__, __LINE__);
             int radio_freq_band = 0;
+	    scan_trigger_counter = 0;
             // copy the bss info to lcb
             memset(&ext->last_connected_bss, 0, sizeof(bss_candidate_t));
             memcpy(&ext->last_connected_bss.external_ap, &sta_data->bss_info, sizeof(wifi_bss_info_t));
@@ -1573,8 +1723,18 @@ int process_ext_sta_conn_status(vap_svc_t *svc, void *arg)
 
             // change the state
             ext_set_conn_state(ext, connection_state_connected, __func__, __LINE__);
-
-            /* Self heal to check if the connected interface received valid ip after a timeout if not trigger a reconnection */
+	    wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD] scan-trigger-counter : %d\n", __func__, __LINE__, scan_trigger_counter);
+	    ret = publish_endpoint_status_to_wan(ctrl, sta_data->stats.connect_status);
+            if (ret == RETURN_ERR) {
+	        wifi_util_dbg_print(WIFI_CTRL,"%s:%d Error in publishing the status\n", __func__, __LINE__);
+	    }
+        
+	    wifi_hal_add_station_bridge(sta_data->interface_name,bridge_name);
+
+            snprintf(cmd, sizeof(cmd), "ip link set dev %s up", bridge_name);
+            wifi_util_dbg_print(WIFI_CTRL,"%s:%d cmd : %s\n",__func__,__LINE__, cmd);
+            get_stubs_descriptor()->v_secure_system_fn(cmd);
+	    /* Self heal to check if the connected interface received valid ip after a timeout if not trigger a reconnection */
 
             if (ext->ext_udhcp_ip_check_id != 0) {
                 scheduler_cancel_timer_task(ctrl->sched, ext->ext_udhcp_ip_check_id);
@@ -1603,7 +1763,9 @@ int process_ext_sta_conn_status(vap_svc_t *svc, void *arg)
             radio_feat = (wifi_radio_feature_param_t *)get_wifidb_radio_feat_map(index);
             radio_params = (wifi_radio_operationParam_t *)get_wifidb_radio_map(index);
             if (radio_params != NULL) {
+                wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD]\n", __func__, __LINE__);
                 if ((radio_params->channel != sta_data->stats.channel) || (radio_params->channelWidth != sta_data->stats.channelWidth)) {
+                    wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD]\n", __func__, __LINE__);
                     pthread_mutex_lock(&mgr->data_cache_lock);
                     radio_params->channel = sta_data->stats.channel;
                     radio_params->channelWidth = sta_data->stats.channelWidth;
@@ -1642,6 +1804,7 @@ int process_ext_sta_conn_status(vap_svc_t *svc, void *arg)
         }
     } else if (sta_data->stats.connect_status == wifi_connection_status_ap_not_found || sta_data->stats.connect_status == wifi_connection_status_disconnected) {
 
+                wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD]\n", __func__, __LINE__);
         apply_pending_channel_change(svc, sta_data->stats.vap_index);
 
         if (ext->conn_state == connection_state_connected &&
@@ -1676,42 +1839,83 @@ int process_ext_sta_conn_status(vap_svc_t *svc, void *arg)
             ext->ext_disconnection_event_timeout_handler_id = 0;
         }
 
-        if (ext->conn_state == connection_state_connection_to_nb_in_progress) {
-            candidate = &ext->new_bss;
+	if (scan_trigger_counter >= scan_trigger_threshold) {
+	    wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD] scan_trigger_counter : %d scan_trigger_threshold : %d\n", __func__, __LINE__, scan_trigger_counter, scan_trigger_threshold);
+            scan_trigger_counter = 0;
+	    ret = publish_endpoint_status_to_wan(ctrl, sta_data->stats.connect_status);
+            if (ret == RETURN_ERR) {
+                wifi_util_dbg_print(WIFI_CTRL,"%s:%d Error in publishing the status\n", __func__, __LINE__);
+            }
+	    //ret = set_endpoint_enable(sta_data->stats.connect_status);
+	}
+
+        //Workaround for sta disconnection
+        wifi_util_dbg_print(WIFI_CTRL,"%s:%d Deleting link\n", __func__, __LINE__);	
+	memset(cmd, '\0', 128);
+	snprintf(cmd, sizeof(cmd), "ovs-vsctl del-port brww0 wl1");
+        wifi_util_dbg_print(WIFI_CTRL,"%s:%d cmd : %s\n",__func__,__LINE__, cmd);
+        get_stubs_descriptor()->v_secure_system_fn(cmd);
+        wifi_util_dbg_print(WIFI_CTRL,"%s:%d Link Deletion done\n", __func__, __LINE__);	
+        
+	if (ext->conn_state == connection_state_connection_to_nb_in_progress) {
+            wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD]\n", __func__, __LINE__);
+	    candidate = &ext->new_bss;
             found_candidate = true;
         } else if ((ext->conn_state == connection_state_connection_to_lcb_in_progress) ||
                 (ext->conn_state == connection_state_connected)) {
 
+            wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD]\n", __func__, __LINE__);
             if (ext->is_radio_ignored == true) {
+                wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD]\n", __func__, __LINE__);
                 candidate = NULL;
                 found_candidate = false;
+		scan_trigger_counter++;
+                wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD] SCAN TRIGGER COUNTER [%d]\n", __func__, __LINE__, scan_trigger_counter);
                 ext_set_conn_state(ext, connection_state_disconnected_scan_list_none, __func__,
                     __LINE__);
             } else {
-                candidate = &ext->last_connected_bss;
+                wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD]\n", __func__, __LINE__);
+		candidate = &ext->last_connected_bss;
                 found_candidate = true;
                 ext_set_conn_state(ext, connection_state_connection_to_lcb_in_progress, __func__,
                     __LINE__);
             }
         } else if (ext->conn_state == connection_state_connection_in_progress) {
+                wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD]\n", __func__, __LINE__);
             candidate = ext->candidates_list.scan_list;
             for (i = 0; i < ext->candidates_list.scan_count; i++) {
+                wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD]\n", __func__, __LINE__);
                 if ((candidate->conn_attempt == connection_attempt_wait) && (candidate->conn_retry_attempt < STA_MAX_CONNECT_ATTEMPT)) {
+                    wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD]\n", __func__, __LINE__);
                     found_candidate = true;
                     break;
                 }
                 candidate++;
+                wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD]\n", __func__, __LINE__);
             }
             if (found_candidate == false) {
+                wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD]\n", __func__, __LINE__);
                 candidate = NULL;
             }
         } else if (ext->conn_state == connection_state_disconnection_in_progress) {
+                wifi_util_info_print(WIFI_CTRL, "%s:%d[PRAMOD]\n", __func__, __LINE__);
             ext_set_conn_state(ext, connection_state_connection_in_progress, __func__, __LINE__);
             candidate = ext->candidates_list.scan_list;
             found_candidate = true;
         }
     }
 
+#if 0	
+    if (send_event == true) {
+        wifi_util_dbg_print(WIFI_CTRL,"Now connected %s:%d\n",__func__,__LINE__);
+	wifi_hal_add_station_bridge(sta_data->interface_name,bridge_name);
+
+        snprintf(cmd, sizeof(cmd), "ip link set dev %s up", bridge_name);
+        wifi_util_dbg_print(WIFI_CTRL,"%s:%d cmd : %s\n",__func__,__LINE__, cmd);
+        get_stubs_descriptor()->v_secure_system_fn(cmd);
+        wifi_util_dbg_print(WIFI_CTRL,"Now connected %s:%d\n",__func__,__LINE__);
+	sprintf(name, "Device.WiFi.STA.%d.Connection.Status", index + 1);
+		
     if (!is_devtype_pod() && send_event == true) {
         sprintf(name, "Device.WiFi.STA.%d.Connection.Status", index + 1);
 
diff --git a/source/core/wifi_ctrl.c b/source/core/wifi_ctrl.c
index 11575463a..b9574b5e4 100644
--- a/source/core/wifi_ctrl.c
+++ b/source/core/wifi_ctrl.c
@@ -258,12 +258,12 @@ void sta_selfheal_handing(wifi_ctrl_t *ctrl, vap_svc_t *l_svc)
             wifi_util_error_print(WIFI_CTRL,"%s:%d selfheal: STA connection failed for %d minutes, publish selfheal connection timeout\n",
                             __func__, __LINE__, selfheal_event_publish_time());
             /* publish selfheal STA Connection Timeout  device */
-            selfheal_event_publish(ctrl);
+            //selfheal_event_publish(ctrl);
             disconnected_time = 0;
             connection_timeout = 0;
         } else if (((disconnected_time * STA_CONN_RETRY_TIMEOUT) >= ((selfheal_event_publish_time() * 60) / 2)) && (radio_reset_triggered == false)) {
-            reset_wifi_radios();
-            radio_reset_triggered = true;
+            //reset_wifi_radios();
+            //radio_reset_triggered = true;
         } else if ((connection_timeout * STA_CONN_RETRY_TIMEOUT) >= MAX_CONNECTION_ALGO_TIMEOUT) {
             l_svc->event_fn(l_svc, wifi_event_type_exec, wifi_event_exec_timeout, vap_svc_event_none, NULL);
             connection_timeout = 0;
@@ -278,11 +278,11 @@ void sta_selfheal_handing(wifi_ctrl_t *ctrl, vap_svc_t *l_svc)
 bool is_sta_enabled(void)
 {
     wifi_ctrl_t *ctrl = (wifi_ctrl_t *)get_wifictrl_obj();
-    //wifi_util_dbg_print(WIFI_CTRL,"[%s:%d] device mode:%d active_gw_check:%d\r\n",
-    //    __func__, __LINE__, ctrl->network_mode, ctrl->active_gw_check);
-    return ((ctrl->network_mode == rdk_dev_mode_type_ext ||
-                ctrl->network_mode == rdk_dev_mode_type_em_node || ctrl->active_gw_check == true) &&
-        ctrl->eth_bh_status == false);
+    wifi_util_dbg_print(WIFI_CTRL,"[%s:%d] device mode:%d active_gw_check:%d and rf_status_down=%d\r\n",
+       __func__, __LINE__, ctrl->network_mode, ctrl->active_gw_check,  ctrl->rf_status_down);
+   return ((ctrl->network_mode == rdk_dev_mode_type_ext ||
+              ctrl->network_mode == rdk_dev_mode_type_em_node || ctrl->active_gw_check == true || 
+              ctrl->rf_status_down == true ) &&  ctrl->eth_bh_status == false);
 }
 
 void ctrl_queue_loop(wifi_ctrl_t *ctrl)
@@ -760,7 +760,7 @@ void start_gateway_vaps()
     vap_svc_t *priv_svc, *pub_svc, *mesh_gw_svc;
     unsigned int value;
     wifi_ctrl_t *ctrl;
-
+    wifi_util_dbg_print(WIFI_CTRL,"%s and %d\n",__func__,__LINE__);
     ctrl = (wifi_ctrl_t *)get_wifictrl_obj();
 
     priv_svc = get_svc_by_type(ctrl, vap_svc_type_private);
@@ -787,10 +787,18 @@ void start_gateway_vaps()
     value = false;
     if (bus_get_active_gw_parameter(WIFI_ACTIVE_GATEWAY_CHECK, &value) == RETURN_OK) {
         ctrl->active_gw_check = value;
-        if (is_sta_enabled() == true) {
+    }
+    if (is_sta_enabled() == true) {
             wifi_util_info_print(WIFI_CTRL, "%s:%d start mesh sta\n",__func__, __LINE__);
             start_extender_vaps();
-        }
+    }
+    value = false;
+    if (bus_get_active_gw_parameter(RF_STATUS_CHECK, &value) == RETURN_OK) {
+        ctrl->rf_status_down = value;
+    }
+    if (is_sta_enabled() == true) {
+            wifi_util_info_print(WIFI_CTRL, "%s:%d start mesh sta\n",__func__, __LINE__);
+            //start_station_vaps();
     }
 }
 
@@ -1748,6 +1756,11 @@ int start_wifi_ctrl(wifi_ctrl_t *ctrl)
     webconfig_send_full_associate_status(ctrl);
     ctrl->exit_ctrl = false;
     ctrl->ctrl_initialized = true;
+    wifi_util_info_print(WIFI_CTRL,"%s:%d Testing\n", __func__, __LINE__);
+    register_endpoint_components(ctrl);
+    wifi_util_info_print(WIFI_CTRL,"%s:%d Testing\n", __func__, __LINE__);
+    get_stubs_descriptor()->v_secure_system_fn("touch /tmp/wifi_ready_to_process");
+    wifi_util_info_print(WIFI_CTRL,"%s:%d Testing\n", __func__, __LINE__);
     ctrl_queue_loop(ctrl);
 
 #ifdef ONEWIFI_ANALYTICS_APP_SUPPORT
@@ -2155,16 +2168,19 @@ static int bus_check_and_subscribe_events(void* arg)
 
 static int sta_connectivity_selfheal(void* arg)
 {
-    wifi_ctrl_t *ctrl = NULL;
-    vap_svc_t *ext_svc;
+    #if 0
+	wifi_ctrl_t *ctrl = NULL;
+    //vap_svc_t *ext_svc;
 
     ctrl = (wifi_ctrl_t *)get_wifictrl_obj();
     
     ext_svc = get_svc_by_type(ctrl, vap_svc_type_mesh_ext);
     if (is_sta_enabled()) {
         // check sta connectivity selfheal
-        sta_selfheal_handing(ctrl, ext_svc);
+        wifi_util_dbg_print(WIFI_CTRL,"station is enabled and hence selfheal started\n");
+        //sta_selfheal_handing(ctrl, ext_svc);
     }
+	#endif
     return TIMER_TASK_COMPLETE;
 }
 
@@ -2828,7 +2844,6 @@ wifi_radio_capabilities_t *getRadioCapability(UINT radioIndex)
         return NULL;
     }
 
-    wifi_util_dbg_print(WIFI_CTRL, "%s Input radioIndex = %d\n", __FUNCTION__, radioIndex);
 
     return &wifi_hal_cap_obj->wifi_prop.radiocap[radioIndex];
 }
diff --git a/source/core/wifi_ctrl.h b/source/core/wifi_ctrl.h
index 9f784f6f0..8dbe52b99 100644
--- a/source/core/wifi_ctrl.h
+++ b/source/core/wifi_ctrl.h
@@ -264,6 +264,7 @@ typedef struct wifi_ctrl {
     int                 speed_test_running;
     events_bus_data_t   events_bus_data;
     hotspot_cfg_sem_param_t hotspot_sem_param;
+    bool                rf_status_down;
 } wifi_ctrl_t;
 
 
@@ -399,7 +400,7 @@ char *get_assoc_devices_blob();
 void get_subdoc_name_from_vap_index(uint8_t vap_index, int* subdoc);
 int dfs_nop_start_timer(void *args);
 int webconfig_send_full_associate_status(wifi_ctrl_t *ctrl);
-
+void start_station_vaps(bool enable);
 bool hotspot_cfg_sem_wait_duration(uint32_t time_in_sec);
 void hotspot_cfg_sem_signal(bool status);
 
diff --git a/source/core/wifi_ctrl_queue_handlers.c b/source/core/wifi_ctrl_queue_handlers.c
index f0e61e141..d50df509b 100644
--- a/source/core/wifi_ctrl_queue_handlers.c
+++ b/source/core/wifi_ctrl_queue_handlers.c
@@ -3763,6 +3763,11 @@ void handle_webconfig_event(wifi_ctrl_t *ctrl, const char *raw, unsigned int len
             num_ssid += get_list_of_lnf_radius(&mgr->hal_cap.wifi_prop, MAX_NUM_RADIOS,
                 &vap_names[num_ssid]);
             break;
+        case webconfig_subdoc_type_mesh_sta:
+            wifi_util_error_print(WIFI_CTRL, "%s:%d  mesh_sta event pointer\n", __func__, __LINE__);
+            num_ssid += get_list_of_mesh_sta(&mgr->hal_cap.wifi_prop, MAX_NUM_RADIOS,
+                &vap_names[num_ssid]);
+            break;
 
         default:
             break;
diff --git a/source/core/wifi_ctrl_rbus_handlers.c b/source/core/wifi_ctrl_rbus_handlers.c
index 6c374a473..9e9494250 100644
--- a/source/core/wifi_ctrl_rbus_handlers.c
+++ b/source/core/wifi_ctrl_rbus_handlers.c
@@ -107,6 +107,43 @@ static int get_subdoc_type(wifi_provider_response_t *response, webconfig_subdoc_
     return ret;
 }
 
+bus_error_t rf_get_status(char *name, raw_data_t *p_data, bus_user_data_t *user_data)
+{
+    (void)user_data;
+    bus_error_t rc = bus_error_success;
+    wifi_util_dbg_print(WIFI_CTRL,"%s:%d\n",__func__,__LINE__);
+    wifi_ctrl_t *ctrl = (wifi_ctrl_t *)get_wifictrl_obj();
+    if (ctrl == NULL) {
+        wifi_util_error_print(WIFI_CTRL, "%s:%d NULL pointers\n", __func__, __LINE__);
+        return bus_error_general;
+    }
+    p_data->data_type = bus_data_type_boolean;
+    p_data->raw_data.b = ctrl->rf_status_down;
+    return rc;
+}
+
+bus_error_t rf_set_status(char *name, raw_data_t *p_data, bus_user_data_t *user_data)
+{
+    (void)user_data;
+    bus_error_t rc = bus_error_success;
+    bool rf_status = false;
+    wifi_ctrl_t *ctrl = (wifi_ctrl_t *)get_wifictrl_obj();
+    if (ctrl == NULL) {
+        wifi_util_error_print(WIFI_CTRL, "%s:%d NULL pointers\n", __func__, __LINE__);
+        return bus_error_general;
+    }
+    if (p_data->data_type != bus_data_type_boolean) {
+        wifi_util_error_print(WIFI_CTRL, "%s:%d Invalid data input\n", __func__, __LINE__);
+        return bus_error_general;
+    }
+    rf_status = p_data->raw_data.b;
+    ctrl->rf_status_down = rf_status;
+    start_station_vaps(rf_status);
+
+    return rc;
+
+}
+
 int stats_bus_publish(wifi_ctrl_t *ctrl, void *stats_data)
 {
     webconfig_subdoc_data_t *data;
@@ -739,6 +776,54 @@ bus_error_t webconfig_get_dml_subdoc(char *event_name, raw_data_t *p_data, bus_u
     return bus_error_success;
 }
 
+bus_error_t get_endpoint_status(char *event_name, raw_data_t *p_data, bus_user_data_t *user_data)
+{
+    (void)user_data;
+    unsigned int index, vap_index = 0, i = 0, no_of_radio = 0,str_size = 0 ;
+    wifi_mgr_t *mgr = (wifi_mgr_t *)get_wifimgr_obj();
+    wifi_vap_info_map_t *vap_map;
+    bool connected = false;
+    char status[32] = "";
+
+    if (event_name == NULL) {
+        wifi_util_error_print(WIFI_CTRL, "%s:%d property name is not found\r\n", __FUNCTION__,
+            __LINE__);
+        return bus_error_invalid_input;
+    }
+    no_of_radio = getNumberRadios();
+    for(index = 0 ; index < no_of_radio ;index++)
+    {
+        vap_map = &mgr->radio_config[index].vaps.vap_map;
+        vap_index = get_sta_vap_index_for_radio(&mgr->hal_cap.wifi_prop, index);
+        for (i = 0; i < vap_map->num_vaps; i++) {
+            if (vap_map->vap_array[i].vap_index == vap_index &&vap_map->vap_array[i].u.sta_info.conn_status == wifi_connection_status_connected ) {
+                wifi_util_info_print(WIFI_CTRL,"%s:%d connected vap_index=%d\n",__func__,__LINE__,i);
+                connected = true;
+                break;
+            }
+        }
+    }
+    if (connected) {
+        wifi_util_info_print(WIFI_CTRL,"station is connected setting Up\n");
+        snprintf(status,sizeof(status),"Up");
+    } else {
+        wifi_util_info_print(WIFI_CTRL,"station is not connected setting false\n");
+        snprintf(status,sizeof(status),"Down");
+   }
+   str_size = strlen(status) + 1;
+    p_data->data_type = bus_data_type_string;
+    p_data->raw_data.bytes = malloc(str_size);
+    if (p_data->raw_data.bytes == NULL) {
+        wifi_util_error_print(WIFI_CTRL,"%s:%d memory allocation is failed:%d\r\n",__func__,
+            __LINE__, str_size);
+        return bus_error_out_of_resources;
+    }
+    strncpy((char *)p_data->raw_data.bytes, status, str_size);
+    p_data->raw_data_len = str_size;
+
+    return bus_error_success;
+}
+
 bus_error_t webconfig_set_subdoc(char *event_name, raw_data_t *p_data, bus_user_data_t *user_data)
 {
     (void)user_data;
@@ -834,7 +919,6 @@ static void wan_failover_handler(char *event_name, raw_data_t *p_data, void *use
 
     wifi_util_dbg_print(WIFI_CTRL, "%s:%d: recv data:%d\r\n", __func__, __LINE__, data_value);
 }
-
 static void hotspotTunnelHandler(char *event_name, raw_data_t *p_data, void *userData)
 {
     (void)userData;
@@ -3088,6 +3172,33 @@ bus_error_t set_force_vap_apply(char *name, raw_data_t *p_data, bus_user_data_t
     return bus_error_invalid_input;
 }
 
+void register_endpoint_components(wifi_ctrl_t *ctrl)
+{
+    int rc = bus_error_success;
+    int num_elements;
+    wifi_util_info_print(WIFI_CTRL,"%s:%d Testing\n", __func__, __LINE__);
+    bus_data_element_t data_elements[] = {
+                         { WIFI_ENDPOINT_CONNECT_STATUS, bus_element_type_method,
+                                    { get_endpoint_status,NULL, NULL, NULL, NULL, NULL }, slow_speed, ZERO_TABLE,
+                                    { bus_data_type_string, true, 0, 0, 0, NULL } },
+
+			{ RF_STATUS_CHECK, bus_element_type_method,
+                                    { rf_get_status, rf_set_status, NULL, NULL, NULL,NULL }, slow_speed, ZERO_TABLE,
+                                    { bus_data_type_boolean, true, 0, 0, 0, NULL } },
+     };
+     num_elements = (sizeof(data_elements) / sizeof(bus_data_element_t));
+    wifi_util_info_print(WIFI_CTRL,"%s:%d num-elements : %d\n", __func__, __LINE__, num_elements);
+     rc = get_bus_descriptor()->bus_reg_data_element_fn(&ctrl->handle,data_elements, num_elements);
+     wifi_util_error_print(WIFI_CTRL, "%s %d rc : %d\n", __func__, __LINE__, rc);
+     if (rc != bus_error_success) {
+        wifi_util_error_print(WIFI_CTRL, "%s %dbus: bus_regDataElements failed\n", __FUNCTION__, __LINE__);
+        return;
+     }
+      wifi_util_error_print(WIFI_CTRL, "%s %d bus: bus_regDataElements done\n", __FUNCTION__, __LINE__);
+     return;
+}
+
+
 void bus_register_handlers(wifi_ctrl_t *ctrl)
 {
     int rc = bus_error_success;
diff --git a/source/core/wifi_ctrl_webconfig.c b/source/core/wifi_ctrl_webconfig.c
index 135b42705..924e279f0 100644
--- a/source/core/wifi_ctrl_webconfig.c
+++ b/source/core/wifi_ctrl_webconfig.c
@@ -869,6 +869,7 @@ int webconfig_hal_vap_apply_by_name(wifi_ctrl_t *ctrl, webconfig_subdoc_decoded_
 
             start_wifi_sched_timer(vap_info->vap_index, ctrl, wifi_vap_sched);
 
+	    wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d [TESTING]\n", __func__, __LINE__);
             if (svc->update_fn(svc, tgt_radio_idx, p_tgt_vap_map, &tgt_rdk_vap_info) != RETURN_OK) {
                 wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d: failed to apply\n", __func__, __LINE__);
                 memset(update_status, 0, sizeof(update_status));
@@ -900,6 +901,7 @@ int webconfig_hal_vap_apply_by_name(wifi_ctrl_t *ctrl, webconfig_subdoc_decoded_
              * So, it won't be updated within update_fn -> wifidb_update_wifi_vap_info. Thats the reason why I leaved memcpy here. 
              */
             memcpy(mgr_vap_info, &p_tgt_vap_map->vap_array[0], sizeof(wifi_vap_info_t));
+	    wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d [TESTING]\n", __func__, __LINE__);
 
             // This block of code is only used for updating VAP mac.
             //if (vap_info->vap_mode == wifi_vap_mode_ap && is_bssid_valid(p_tgt_vap_map->vap_array[0].u.bss_info.bssid)) {
@@ -2739,6 +2741,78 @@ webconfig_error_t webconfig_ctrl_apply(webconfig_subdoc_t *doc, webconfig_subdoc
     return ((ret == RETURN_OK) ? webconfig_error_none:webconfig_error_apply);
 }
 
+void start_station_vaps(bool rf_status)
+{
+    wifi_util_error_print(WIFI_CTRL, "%s:%d \n", __func__, __LINE__);
+    webconfig_subdoc_data_t *data = NULL;
+    int status = RETURN_OK;
+    int vap_index , radio_index = 0, vap_array_index = 0, band = 0;
+    char *str;
+    char password[128] = {0};
+    wifi_vap_name_t vap_names[MAX_NUM_RADIOS] = {0};
+    wifi_ctrl_t *ctrl;
+    ctrl = (wifi_ctrl_t *)get_wifictrl_obj();
+
+    data = (webconfig_subdoc_data_t *) malloc(sizeof(webconfig_subdoc_data_t));
+    if (data == NULL) {
+        wifi_util_error_print(WIFI_CTRL, "%s: malloc failed to allocate webconfig_subdoc_data_t, size %d\n", \
+                              __func__, sizeof(webconfig_subdoc_data_t));
+        return;
+    }
+
+    webconfig_init_subdoc_data(data);
+    unsigned int num_vaps = get_list_of_mesh_sta(&data->u.decoded.hal_cap.wifi_prop, MAX_NUM_RADIOS, &vap_names[0]);
+
+    for (size_t i = 0; i < num_vaps; i++) {
+        wifi_util_error_print(WIFI_CTRL, "%s:%d \n", __func__, __LINE__);
+        vap_index = convert_vap_name_to_index(&data->u.decoded.hal_cap.wifi_prop,vap_names[i]);
+        if (vap_index == RETURN_ERR) {
+            continue;
+        }
+        status = get_vap_and_radio_index_from_vap_instance(&data->u.decoded.hal_cap.wifi_prop, vap_index, (uint8_t *)&radio_index, (uint8_t *)&vap_array_index);
+        if (status == RETURN_ERR) {
+            break;
+        } else {
+            convert_radio_index_to_freq_band(&data->u.decoded.hal_cap.wifi_prop, radio_index, &band);
+            if (rf_status) {
+                wifi_util_error_print(WIFI_CTRL, "%s:%d rf_status=%d \n", __func__, __LINE__,rf_status);
+                snprintf(data->u.decoded.radios[radio_index].vaps.vap_map.vap_array[vap_array_index].u.sta_info.ssid,sizeof(data->u.decoded.radios[radio_index].vaps.vap_map.vap_array[vap_array_index].u.sta_info.ssid),"Xfinity_Ignite");
+                if (band == WIFI_FREQUENCY_6_BAND) {
+                    data->u.decoded.radios[radio_index].vaps.vap_map.vap_array[vap_array_index].u.sta_info.security.mode = wifi_security_mode_wpa3_enterprise;
+                } else {
+                    data->u.decoded.radios[radio_index].vaps.vap_map.vap_array[vap_array_index].u.sta_info.security.mode = wifi_security_mode_wpa2_enterprise;
+                }
+
+            memset(&data->u.decoded.radios[radio_index].vaps.vap_map.vap_array[vap_array_index].u.sta_info.security.u.radius.ip, 0, sizeof(data->u.decoded.radios[radio_index].vaps.vap_map.vap_array[vap_array_index].u.sta_info.security.u.radius.ip));
+            memset(&data->u.decoded.radios[radio_index].vaps.vap_map.vap_array[vap_array_index].u.sta_info.security.u.radius.s_ip, 0, sizeof(data->u.decoded.radios[radio_index].vaps.vap_map.vap_array[vap_array_index].u.sta_info.security.u.radius.s_ip));
+            } else {
+                wifi_util_error_print(WIFI_CTRL, "%s:%d rf_status=%d \n", __func__, __LINE__,rf_status);
+                snprintf(data->u.decoded.radios[radio_index].vaps.vap_map.vap_array[vap_array_index].u.sta_info.ssid,sizeof(data->u.decoded.radios[radio_index].vaps.vap_map.vap_array[vap_array_index].u.sta_info.ssid),"we.connect.yellowstone");
+                if (band == WIFI_FREQUENCY_6_BAND) {
+                    data->u.decoded.radios[radio_index].vaps.vap_map.vap_array[vap_array_index].u.sta_info.security.mode = wifi_security_mode_wpa3_personal;
+                } else {
+                    data->u.decoded.radios[radio_index].vaps.vap_map.vap_array[vap_array_index].u.sta_info.security.mode = wifi_security_mode_wpa2_personal;
+                }
+                memset(password, 0, sizeof(password));
+               if (wifi_hal_get_default_keypassphrase(password,vap_index) == 0) {
+                    strcpy(data->u.decoded.radios[radio_index].vaps.vap_map.vap_array[vap_array_index].u.sta_info.security.u.key.key, password);
+                } else {
+                    strcpy(data->u.decoded.radios[radio_index].vaps.vap_map.vap_array[vap_array_index].u.sta_info.security.u.key.key, "12345678");
+                }
+            }
+
+        }
+    }
+
+    if (webconfig_encode(&ctrl->webconfig, data, webconfig_subdoc_type_mesh_sta) == webconfig_error_none) {
+        wifi_util_error_print(WIFI_CTRL, "%s:%d -webconfig_encode success\n", __FUNCTION__, __LINE__);
+        str = data->u.encoded.raw;
+        push_event_to_ctrl_queue(str, strlen(str), wifi_event_type_webconfig, wifi_event_webconfig_set_data_dml, NULL);
+
+     } else {
+        webconfig_data_free(data);
+     }
+}
 // register subdocs with webconfig_framework
 int register_with_webconfig_framework()
 {
diff --git a/source/db/wifi_db.c b/source/db/wifi_db.c
index 5ac7029bd..989c5a9f9 100644
--- a/source/db/wifi_db.c
+++ b/source/db/wifi_db.c
@@ -313,8 +313,7 @@ static int init_vap_config_default(int vap_index, wifi_vap_info_t *config,
             wifi_util_error_print(WIFI_DB, "%s:%d: Incorrect password length %d for vap '%s'\n", __func__, __LINE__, strlen(cfg.u.sta_info.security.u.key.key), vap_name);
             strncpy(cfg.u.sta_info.security.u.key.key, INVALID_KEY, sizeof(cfg.u.sta_info.security.u.key.key));
         }
-
-        cfg.u.sta_info.scan_params.channel.band = band;
+	cfg.u.sta_info.scan_params.channel.band = band;
         cfg.u.sta_info.scan_params.channel.channel = 0;
         cfg.u.sta_info.conn_status = wifi_connection_status_disabled;
         memset(&cfg.u.sta_info.bssid, 0, sizeof(cfg.u.sta_info.bssid));
diff --git a/source/db/wifi_db_apis.c b/source/db/wifi_db_apis.c
index db0a30c94..ee0177503 100644
--- a/source/db/wifi_db_apis.c
+++ b/source/db/wifi_db_apis.c
@@ -8981,7 +8981,6 @@ int wifi_mgr_bus_subsription(bus_handle_t *handle)
 {
     int rc;
     char *component_name = "WifiMgr";
-
     rc = get_bus_descriptor()->bus_open_fn(handle, component_name);
     if (rc != bus_error_success) {
         wifi_util_error_print(WIFI_MGR, "%s:%d bus: bus_open_fn open failed for component:%s, rc:%d\n",
diff --git a/source/utils/wifi_util.c b/source/utils/wifi_util.c
index 112d48ec4..a5f19b74e 100644
--- a/source/utils/wifi_util.c
+++ b/source/utils/wifi_util.c
@@ -3907,11 +3907,23 @@ bool is_vap_param_config_changed(wifi_vap_info_t *vap_info_old, wifi_vap_info_t
     if (isSta) {
         // Ignore change of conn_status, scan_params, mac to avoid reconfiguration and disconnection
         // BSSID change is handled by event.
-        if (IS_STR_CHANGED(vap_info_old->u.sta_info.ssid, vap_info_new->u.sta_info.ssid,
+	wifi_ctrl_t *ctrl = (wifi_ctrl_t *)get_wifictrl_obj();
+	wifi_util_error_print(WIFI_WEBCONFIG, "[%s %d] rf-status : %d Old SSID : %s New SSID : %s\n", __func__, __LINE__, ctrl->rf_status_down, vap_info_old->u.sta_info.ssid, vap_info_new->u.sta_info.ssid);
+        if ((ctrl->rf_status_down == 1) && ((strcmp(vap_info_old->u.sta_info.ssid, "Xfinity_Ignite") == 0) && (strcmp(vap_info_new->u.sta_info.ssid, "Xfinity_Ignite") == 0))) {
+	    wifi_util_error_print(WIFI_WEBCONFIG, "[%s %d] rf-status : %d Old SSID : %s New SSID : %s\n", __func__, __LINE__, ctrl->rf_status_down, vap_info_old->u.sta_info.ssid, vap_info_new->u.sta_info.ssid);
+	    wifi_util_error_print(WIFI_WEBCONFIG, "[%s %d] Docsis disabled and SSID same. Overriding the update\n", __func__, __LINE__);
+	    return true;
+	}
+	if (IS_STR_CHANGED(vap_info_old->u.sta_info.ssid, vap_info_new->u.sta_info.ssid,
                 sizeof(ssid_t)) ||
             IS_CHANGED(vap_info_old->u.sta_info.enabled, vap_info_new->u.sta_info.enabled) ||
             IS_BIN_CHANGED(&vap_info_old->u.sta_info.security, &vap_info_new->u.sta_info.security,
                 sizeof(wifi_vap_security_t))) {
+	        wifi_util_error_print(WIFI_WEBCONFIG, "[%s %d] rf-status : %d Old SSID : %s New SSID : %s\n", __func__, __LINE__, ctrl->rf_status_down, vap_info_old->u.sta_info.ssid, vap_info_new->u.sta_info.ssid);
+		if ((ctrl->rf_status_down == 1) && ((strcmp(vap_info_old->u.sta_info.ssid, "Xfinity_Ignite") == 0) && (strcmp(vap_info_new->u.sta_info.ssid, "we.connect.yellowstone") == 0))) {
+		    wifi_util_error_print(WIFI_WEBCONFIG, "[%s %d] Device is in docsis disabled mode. Avoid applying blob for mesh sta\n", __func__, __LINE__);
+		    return false;
+		}
             return true;
         }
     } else {
diff --git a/source/webconfig/wifi_ovsdb_translator.c b/source/webconfig/wifi_ovsdb_translator.c
index 9cbd05f21..2c10a554d 100644
--- a/source/webconfig/wifi_ovsdb_translator.c
+++ b/source/webconfig/wifi_ovsdb_translator.c
@@ -1944,6 +1944,46 @@ webconfig_error_t translate_radio_object_to_ovsdb_radio_state_for_dml(webconfig_
     return webconfig_error_none;
 }
 
+BOOL update_secmode_for_wpa3_sta(wifi_vap_info_t *vap_info, char *mode_str, int mode_len, char *encrypt_str, int encrypt_len, bool to_ovsdb)
+{
+    int ret = false;
+    if (vap_info ==  NULL) {
+        wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d NULL vap_info Pointer\n", __func__, __LINE__);
+        return ret;
+    }
+
+     wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d to-ovsdb : %d sec-mode : %d\n", __func__, __LINE__, to_ovsdb, vap_info->u.sta_info.security.mode);
+    if (to_ovsdb) {
+        if ((vap_info->u.sta_info.security.mode == wifi_security_mode_wpa3_transition) || (vap_info->u.sta_info.security.mode == wifi_security_mode_wpa3_personal) || (vap_info->u.sta_info.security.mode == wifi_security_mode_wpa3_compatibility)) {
+            
+	    wifi_util_dbg_print(WIFI_WEBCONFIG, "%s:%d[TESTING]\n", __func__, __LINE__);
+	    snprintf(mode_str, mode_len, "2");
+            snprintf(encrypt_str, encrypt_len, "WPA-PSK");
+            ret = true;
+        }
+        else if (vap_info->u.sta_info.security.mode == wifi_security_mode_wpa3_enterprise) {
+	    wifi_util_dbg_print(WIFI_WEBCONFIG, "%s:%d[TESTING]\n", __func__, __LINE__);
+            snprintf(mode_str, mode_len, "2");
+            snprintf(encrypt_str, encrypt_len, "WPA-EAP");
+            ret = true;
+        }
+        else if (vap_info->u.sta_info.security.mode == wifi_security_mode_enhanced_open) {
+	    wifi_util_dbg_print(WIFI_WEBCONFIG, "%s:%d[TESTING]\n", __func__, __LINE__);
+            snprintf(encrypt_str, encrypt_len, "OPEN");
+            ret = true;
+        }
+    } else {
+        if ((vap_info->u.sta_info.security.mode == wifi_security_mode_wpa3_transition) || (vap_info->u.sta_info.security.mode == wifi_security_mode_wpa3_personal)
+        || (vap_info->u.sta_info.security.mode == wifi_security_mode_wpa3_compatibility)
+        || (vap_info->u.sta_info.security.mode == wifi_security_mode_wpa3_enterprise) || (vap_info->u.sta_info.security.mode == wifi_security_mode_enhanced_open)) {
+	    wifi_util_dbg_print(WIFI_WEBCONFIG, "%s:%d[TESTING]\n", __func__, __LINE__);
+            ret = true;
+        }
+    }
+    wifi_util_dbg_print(WIFI_WEBCONFIG, "%s:%d[TESTING] ret : %d\n", __func__, __LINE__, ret);
+    return ret;
+}
+
 BOOL update_secmode_for_wpa3(wifi_vap_info_t *vap_info, char *mode_str, int mode_len, char *encrypt_str, int encrypt_len, bool to_ovsdb)
 {
     int ret = false;
@@ -1951,19 +1991,24 @@ BOOL update_secmode_for_wpa3(wifi_vap_info_t *vap_info, char *mode_str, int mode
         wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d NULL vap_info Pointer\n", __func__, __LINE__);
         return ret;
     }
+    wifi_util_dbg_print(WIFI_WEBCONFIG, "%s:%d[TESTING]\n", __func__, __LINE__);
+     wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d to-ovsdb : %d sec-mode : %d\n", __func__, __LINE__, to_ovsdb, vap_info->u.sta_info.security.mode);
 
     if (to_ovsdb) {
         if ((vap_info->u.bss_info.security.mode == wifi_security_mode_wpa3_transition) || (vap_info->u.bss_info.security.mode == wifi_security_mode_wpa3_personal) || (vap_info->u.bss_info.security.mode == wifi_security_mode_wpa3_compatibility)) {
+	    wifi_util_dbg_print(WIFI_WEBCONFIG, "%s:%d[TESTING]\n", __func__, __LINE__);
             snprintf(mode_str, mode_len, "2");
             snprintf(encrypt_str, encrypt_len, "WPA-PSK");
             ret = true;
         }
         else if (vap_info->u.bss_info.security.mode == wifi_security_mode_wpa3_enterprise) {
+	    wifi_util_dbg_print(WIFI_WEBCONFIG, "%s:%d[TESTING]\n", __func__, __LINE__);
             snprintf(mode_str, mode_len, "2");
             snprintf(encrypt_str, encrypt_len, "WPA-EAP");
             ret = true;
         }
         else if (vap_info->u.bss_info.security.mode == wifi_security_mode_enhanced_open) {
+	    wifi_util_dbg_print(WIFI_WEBCONFIG, "%s:%d[TESTING]\n", __func__, __LINE__);
             snprintf(encrypt_str, encrypt_len, "OPEN");
             ret = true;
         }
@@ -1971,9 +2016,11 @@ BOOL update_secmode_for_wpa3(wifi_vap_info_t *vap_info, char *mode_str, int mode
         if ((vap_info->u.bss_info.security.mode == wifi_security_mode_wpa3_transition) || (vap_info->u.bss_info.security.mode == wifi_security_mode_wpa3_personal)
         || (vap_info->u.bss_info.security.mode == wifi_security_mode_wpa3_compatibility)
         || (vap_info->u.bss_info.security.mode == wifi_security_mode_wpa3_enterprise) || (vap_info->u.bss_info.security.mode == wifi_security_mode_enhanced_open)) {
+	    wifi_util_dbg_print(WIFI_WEBCONFIG, "%s:%d[TESTING]\n", __func__, __LINE__);
             ret = true;
         }
     }
+    wifi_util_dbg_print(WIFI_WEBCONFIG, "%s:%d[TESTING] ret : %d\n", __func__, __LINE__, ret);
     return ret;
 }
 
@@ -2380,13 +2427,15 @@ webconfig_error_t translate_sta_vap_info_to_ovsdb_config_personal_sec(const wifi
 
             memset(str_mode, 0, sizeof(str_mode));
             memset(str_encryp, 0, sizeof(str_encryp));
-            if (!update_secmode_for_wpa3((wifi_vap_info_t *)vap, str_mode, sizeof(str_mode), str_encryp, sizeof(str_encryp), true)) {
+            if (!update_secmode_for_wpa3_sta((wifi_vap_info_t *)vap, str_mode, sizeof(str_mode), str_encryp, sizeof(str_encryp), true)) {
+                wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d\n", __func__, __LINE__);
                 wifi_security_modes_t mode_enum = vap->u.sta_info.security.mode;
                 wifi_encryption_method_t encryp_enum = vap->u.sta_info.security.encr;
                 if ((key_mgmt_conversion_legacy(&mode_enum, &encryp_enum, str_mode, sizeof(str_mode), str_encryp, sizeof(str_encryp), ENUM_TO_STRING)) != RETURN_OK) {
                     wifi_util_error_print(WIFI_WEBCONFIG,"%s:%d: key mgmt conversion failed. security mode 0x%x encr 0x%x\n", __func__, __LINE__, vap->u.sta_info.security.mode, vap->u.sta_info.security.encr);
                     return webconfig_error_translate_to_ovsdb;
                 }
+                wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d\n", __func__, __LINE__);
             }
 
             set_translator_config_security_key_value(vap_row, &sec_index, "encryption", str_encryp);
@@ -2397,13 +2446,16 @@ webconfig_error_t translate_sta_vap_info_to_ovsdb_config_personal_sec(const wifi
 
         } else {
             set_translator_config_security_key_value(vap_row, &sec_index, "encryption", "OPEN");
+            wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d\n", __func__, __LINE__);
         }
     } else {
+        wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d\n", __func__, __LINE__);
         if (vap->u.sta_info.security.mode == wifi_security_mode_none) {
             vap_row->wpa = false;
         } else {
             int len = 0, wpa_psk_index = 0;
             wifi_security_modes_t enum_sec = vap->u.sta_info.security.mode;
+            wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d\n", __func__, __LINE__);
             if ((key_mgmt_conversion(&enum_sec, &len, ENUM_TO_STRING, 0, (char(*)[])vap_row->wpa_key_mgmt)) != RETURN_OK) {
                 wifi_util_error_print(WIFI_WEBCONFIG,"%s:%d: key mgmt conversion failed. security mode 0x%x\n",
                     __func__, __LINE__, vap->u.sta_info.security.mode);
@@ -2412,12 +2464,15 @@ webconfig_error_t translate_sta_vap_info_to_ovsdb_config_personal_sec(const wifi
 
             vap_row->wpa = true;
             vap_row->wpa_key_mgmt_len = len;
+            wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d\n", __func__, __LINE__);
 
             if ((strlen(vap->u.sta_info.security.u.key.key) < MIN_PWD_LEN) || (strlen(vap->u.sta_info.security.u.key.key) > MAX_PWD_LEN)) {
                 wifi_util_error_print(WIFI_WEBCONFIG,"%s:%d: Invalid password length %d\n", __func__, __LINE__, strlen(vap->u.sta_info.security.u.key.key));
                 return webconfig_error_translate_to_ovsdb;
             }
+            wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d\n", __func__, __LINE__);
             set_translator_config_wpa_psks(vap_row, &wpa_psk_index, "key--1", vap->u.sta_info.security.u.key.key);
+            wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d\n", __func__, __LINE__);
         }
     }
 
@@ -3110,7 +3165,8 @@ webconfig_error_t translate_sta_vap_info_to_ovsdb_state_personal_sec(const wifi_
 
             memset(str_mode, 0, sizeof(str_mode));
             memset(str_encryp, 0, sizeof(str_encryp));
-            if (!update_secmode_for_wpa3((wifi_vap_info_t *)vap, str_mode, sizeof(str_mode), str_encryp, sizeof(str_encryp), true)) {
+            if (!update_secmode_for_wpa3_sta((wifi_vap_info_t *)vap, str_mode, sizeof(str_mode), str_encryp, sizeof(str_encryp), true)) {
+                wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d\n", __func__, __LINE__);
                 wifi_security_modes_t mode_enum = vap->u.sta_info.security.mode;
                 wifi_encryption_method_t encryp_enum = vap->u.sta_info.security.encr;
                 if ((key_mgmt_conversion_legacy(&mode_enum, &encryp_enum, 
@@ -3118,8 +3174,10 @@ webconfig_error_t translate_sta_vap_info_to_ovsdb_state_personal_sec(const wifi_
                     wifi_util_dbg_print(WIFI_WEBCONFIG,"%s:%d: key mgmt conversion failed\n", __func__, __LINE__);
                     return webconfig_error_translate_to_ovsdb;
                 }
+                wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d\n", __func__, __LINE__);
             }
 
+            wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d\n", __func__, __LINE__);
             set_translator_state_security_key_value(vap_row, &sec_index, "encryption", str_encryp);
             set_translator_state_security_key_value(vap_row, &sec_index, "mode", str_mode);
             set_translator_state_security_key_value(vap_row, &sec_index, "key", vap->u.sta_info.security.u.key.key);
@@ -3128,17 +3186,20 @@ webconfig_error_t translate_sta_vap_info_to_ovsdb_state_personal_sec(const wifi_
             set_translator_state_security_key_value(vap_row, &sec_index, "encryption", "OPEN");
         }
     } else {
+                wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d\n", __func__, __LINE__);
         if (vap->u.sta_info.security.mode == wifi_security_mode_none) {
             vap_row->wpa = false;
         } else {
             int len = 0, wpa_psk_index = 0;
             wifi_security_modes_t enum_sec = vap->u.sta_info.security.mode;
 
+            wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d\n", __func__, __LINE__);
             if ((key_mgmt_conversion(&enum_sec, &len, ENUM_TO_STRING, 0, (char(*)[])vap_row->wpa_key_mgmt)) != RETURN_OK) {
                 wifi_util_dbg_print(WIFI_WEBCONFIG,"%s:%d: key mgmt conversion failed\n", __func__, __LINE__);
                 return webconfig_error_translate_to_ovsdb;
             }
 
+            wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d\n", __func__, __LINE__);
             vap_row->wpa = true;
             vap_row->wpa_key_mgmt_len = len;
 
@@ -3147,6 +3208,7 @@ webconfig_error_t translate_sta_vap_info_to_ovsdb_state_personal_sec(const wifi_
                 return webconfig_error_translate_to_ovsdb;
             }
             set_translator_state_wpa_psks(vap_row, &wpa_psk_index, "key--1", vap->u.sta_info.security.u.key.key);
+            wifi_util_error_print(WIFI_WEBCONFIG, "%s:%d\n", __func__, __LINE__);
         }
     }
 
